src/main.zig

```diff
@@ -466,9 +469,10 @@ fn processSource(
     const prev_errors = comp.diag.errors;
     comp.renderErrors();
 
-    if (comp.diag.errors != prev_errors) {
-        if (fast_exit) exitWithCleanup(link_objects.items, temp_file_count.*, 1);
-        return; // do not compile if there were errors
+    if (comp.diag.errors != prev_errors or comp.only_syntax) {
+        const has_errors = comp.diag.errors != prev_errors;
+        if (fast_exit) exitWithCleanup(link_objects.items, temp_file_count.*, @boolToInt(has_errors));
+        return; // do not compile if there were errors or we are only checking syntax
```

Comment:
I'd prefer to keep these branches separate for simplicity:
```suggestion
    if (comp.diag.errors != prev_errors) {
        if (fast_exit) exitWithCleanup(link_objects.items, temp_file_count.*, 1);
        return; // do not compile if there were errors
    }

    if (comp.only_syntax) {
        if (fast_exit) std.process.exit(0); // Not linking, no need for cleanup.
        return;
```